File: matrix_generator.cpp
// matrix_generator.cpp
#include "matrix_generator.h"
#include <type_traits>
#include <limits>
#include <random>
#include <algorithm>

namespace mg {

template<typename T>
Matrix<T> generate_dense_block(int rows, int cols, std::mt19937 &rng) {
    Matrix<T> block(rows, std::vector<T>(cols));
    if constexpr (std::is_integral_v<T>) {
        // avoid full int64 range for portability; clamp to -1000..1000 for ints
        std::uniform_int_distribution<long long> dist(-1000, 1000);
        for (int i = 0; i < rows; ++i)
            for (int j = 0; j < cols; ++j)
                block[i][j] = static_cast<T>(dist(rng));
    } else {
        std::uniform_real_distribution<double> dist(-1.0, 1.0);
        for (int i = 0; i < rows; ++i)
            for (int j = 0; j < cols; ++j)
                block[i][j] = static_cast<T>(dist(rng));
    }
    return block;
}

template<typename T>
Matrix<T> generate_matrix(int num_rows,
                          int num_cols,
                          double sparsity,
                          const std::string &pattern,
                          int blocksize,
                          unsigned int seed) {
    Matrix<T> mat(num_rows, std::vector<T>(num_cols));
    std::mt19937 rng(seed);

    auto fill_block = [&](int bi, int bj) {
        int start_row = bi * blocksize;
        int start_col = bj * blocksize;
        int end_row = std::min(start_row + blocksize, num_rows);
        int end_col = std::min(start_col + blocksize, num_cols);
        auto block = generate_dense_block<T>(end_row - start_row, end_col - start_col, rng);
        for (int i = start_row; i < end_row; ++i)
            for (int j = start_col; j < end_col; ++j)
                mat[i][j] = block[i - start_row][j - start_col];
    };

    if (pattern == "random") {
        // Fill randomly with probability of being non-zero = 1 - sparsity
        std::bernoulli_distribution keep(1.0 - sparsity);
        for (int i = 0; i < num_rows; ++i) {
            for (int j = 0; j < num_cols; ++j) {
                if (keep(rng)) {
                    if constexpr (std::is_integral_v<T>) {
                        std::uniform_int_distribution<int> d(-1000, 1000);
                        mat[i][j] = static_cast<T>(d(rng));
                    } else {
                        std::uniform_real_distribution<double> d(-1.0, 1.0);
                        mat[i][j] = static_cast<T>(d(rng));
                    }
                } else {
                    mat[i][j] = static_cast<T>(0);
                }
            }
        }

    } else if (pattern == "checkerboard") {
        int nbi = (num_rows + blocksize - 1) / blocksize;
        int nbj = (num_cols + blocksize - 1) / blocksize;
        for (int i = 0; i < nbi; ++i)
            for (int j = 0; j < nbj; ++j)
                if (((i % 2) ^ (j % 2)) == 0)
                    fill_block(i, j);

    } else if (pattern == "diagonal") {
        int minrc = std::min(num_rows, num_cols);
        for (int i = 0; i < minrc; ++i) {
            auto block = generate_dense_block<T>(1, 1, rng);
            mat[i][i] = block[0][0];
        }

    } else if (pattern == "blockdiagonal") {
        int n = std::min(num_rows, num_cols) / blocksize + 1;
        for (int i = 0; i < n; ++i) fill_block(i, i);

    } else if (pattern == "blockrandom") {
        int nbi = (num_rows + blocksize - 1) / blocksize;
        int nbj = (num_cols + blocksize - 1) / blocksize;
        std::bernoulli_distribution keep(1.0 - sparsity);
        for (int i = 0; i < nbi; ++i)
            for (int j = 0; j < nbj; ++j)
                if (keep(rng)) fill_block(i, j);

    } else {
        throw std::invalid_argument("Unknown pattern: " + pattern);
    }

    return mat;
}

// Explicit template instantiations for common types (to avoid linker issues)
template Matrix<float> generate_matrix<float>(int, int, double, const std::string&, int, unsigned int);
template Matrix<double> generate_matrix<double>(int, int, double, const std::string&, int, unsigned int);
template Matrix<int> generate_matrix<int>(int, int, double, const std::string&, int, unsigned int);
template Matrix<long long> generate_matrix<long long>(int, int, double, const std::string&, int, unsigned int);

template Matrix<float> generate_dense_block<float>(int, int, std::mt19937 &);
template Matrix<double> generate_dense_block<double>(int, int, std::mt19937 &);
template Matrix<int> generate_dense_block<int>(int, int, std::mt19937 &);
template Matrix<long long> generate_dense_block<long long>(int, int, std::mt19937 &);

} // namespace mg

File: matrix_generator.py
# !/usr/bin/python3

import argparse
import numpy as np
from scipy import sparse
from scipy.io import mmwrite
import seaborn as sns
import matplotlib.pyplot as plt

def generate_dense_matrix(block_shape, dtype):
    if dtype.startswith('int'):
        return np.random.randint(np.iinfo(dtype).min, np.iinfo(dtype).max,
                                 block_shape).astype(dtype)
    else:
        return np.random.random(block_shape).astype(dtype)


def generate_matrix(output: str, num_rows: int, num_cols: int, sparsity: float,
                    dtype: str, heatmap: bool = False, pattern: str = 'random',
                    blocksize: int = 16, block_rows: int = None, block_cols: int = None):

    matrix = np.zeros((num_rows, num_cols), dtype=dtype)

    def fill_block(block_i, block_j):
        # choose block sizes: priority block_rows/block_cols if provided,
        # otherwise square blocksize is used for both dimensions
        br = block_rows if block_rows is not None else blocksize
        bc = block_cols if block_cols is not None else blocksize
        # fill the matrix from i to i + br and j to j + bc
        start_row = block_i * br
        start_col = block_j * bc
        end_row = min(start_row + br, num_rows)
        end_col = min(start_col + bc, num_cols)

        block_shape = (end_row - start_row, end_col - start_col)

        matrix[start_row:end_row, start_col:end_col] = generate_dense_matrix(
            block_shape, dtype)

    if pattern == 'random':
        matrix = sparse.random(num_rows, num_cols, (1.0 - sparsity),
                               dtype=dtype if dtype != 'float16' else 'float32'
                               ).toarray().astype(dtype)
    elif pattern == 'checkerboard':
        br = block_rows if block_rows is not None else blocksize
        bc = block_cols if block_cols is not None else blocksize
        num_block_rows = (num_rows + br - 1) // br
        num_block_cols = (num_cols + bc - 1) // bc
        for i in range(num_block_rows):
            for j in range(num_block_cols):
                if (i % 2) ^ (j % 2) == 0:
                    fill_block(i, j)

    elif pattern == 'diagonal':
        for i in range(min(num_rows, num_cols)):
            matrix[i:i+1, i:i+1] = generate_dense_matrix((1,1), dtype)

    elif pattern == 'blockdiagonal':
        br = block_rows if block_rows is not None else blocksize
        bc = block_cols if block_cols is not None else blocksize
        num_block_diag = min(num_rows, num_cols) // max(br, bc) + 1
        for i in range(num_block_diag):
            fill_block(i, i)

    elif pattern == 'blockrandom':
        br = block_rows if block_rows is not None else blocksize
        bc = block_cols if block_cols is not None else blocksize
        num_block_rows = (num_rows + br - 1) // br
        num_block_cols = (num_cols + bc - 1) // bc
        mask = sparse.random(num_block_rows,
                             num_block_cols,
                             (1.0 - sparsity), dtype='int32').toarray()
        for i in range(num_block_rows):
            for j in range(num_block_cols):
                if mask[i][j] != 0:
                    fill_block(i, j)

    elif pattern in ('largerandom', 'largerandom_block', 'largerandom block'):
        # Generate rectangular block-random pattern where blocks are 64 rows x 16 cols
        brow, bcol = 64, 16
        mask = sparse.random(num_rows // brow + 1,
                             num_cols // bcol + 1,
                             (1.0 - sparsity), dtype='int32').toarray()
        for i in range(num_rows // brow + 1):
            for j in range(num_cols // bcol + 1):
                if mask[i][j] != 0:
                    start_row = i * brow
                    start_col = j * bcol
                    end_row = min(start_row + brow, num_rows)
                    end_col = min(start_col + bcol, num_cols)
                    block_shape = (end_row - start_row, end_col - start_col)
                    matrix[start_row:end_row, start_col:end_col] = generate_dense_matrix(block_shape, dtype)


    else:
        print("Pattern not recognized")
        exit(1)

    # If the output filename ends with .mtx, write as Matrix Market sparse file.
    if output is not None and output.lower().endswith('.mtx'):
        # Convert to a sparse COO matrix to store only nonzeros
        # Ensure numeric dtype is supported (promote float16 to float32)
        if hasattr(matrix, 'dtype') and str(matrix.dtype) == 'float16':
            mat_to_write = matrix.astype('float32')
        else:
            mat_to_write = matrix
        mat_sparse = sparse.coo_matrix(mat_to_write)
        mmwrite(output, mat_sparse)
    else:
        matrix_str = ""

        for i in range(num_rows):
            for j in range(num_cols):
                matrix_str += str(matrix[i][j]) + ' '
            matrix_str += '\n'

        if output is not None:
            with open(output, 'w') as f:
                f.write(f"{num_rows} {num_cols}\n")
                f.write(matrix_str)
        else:
            print(matrix_str)

    if heatmap is not None and heatmap:
        sns.heatmap(matrix, cbar=True, cmap=plt.get_cmap('jet'))
        plt.title("Matrix heatmap")

        if output is not None:
            plt.savefig(output + ".png")
            plt.close()
        else:
            plt.show()


def main():
    parser = argparse.ArgumentParser(
        prog='Matrix generator',
        description='Generates matrices with different properties')

    parser.add_argument('-o', '--output',
                        help='Output file. Print to stdout if not specified')
    parser.add_argument('-n', '--nrows', type=int, required=True,
                        help='Number of rows in the matrix')
    parser.add_argument('-m', '--ncols', type=int,
                        help='Number of columns in the matrix. Use nrows if '
                             'not specified')
    parser.add_argument('-s', '--sparsity', type=float, default=0.7,
                        help='Sparsity level of the matrix')
    parser.add_argument('-t', '--type', default='float32',
                        choices=['float16', 'float32', 'int32', 'int64'],
                        help='Type of the values of matrix')
    parser.add_argument('-H', '--heatmap', action='store_true',
                        help='Generate the matrix heatmap')
    parser.add_argument('-p', '--pattern', default='random',
                        choices=['random', 'checkerboard', 'diagonal',
                                 'blockdiagonal', 'blockrandom', 'largerandom',
                                 'r', 'c', 'd', 'D', 'R'],
                        help='Pattern used to fill the matrix')
    parser.add_argument('-b', '--blocksize', default=32, type=int,
                        help='Blocksize of the matrix')
    parser.add_argument('-br', '--blockrows', default=None, type=int,
                        help='Block row size (height) of the matrix')
    parser.add_argument('-bc', '--blockcols', default=None, type=int,
                        help='Block column size (width) of the matrix')

    args = parser.parse_args()
    if args.ncols is None:
        args.ncols = args.nrows

    if args.pattern is None or args.pattern == 'r':
        args.pattern = 'random'
    elif args.pattern == 'c':
        args.pattern = 'checkerboard'
    elif args.pattern == 'd':
        args.pattern = 'diagonal'
    elif args.pattern == 'D':
        args.pattern = 'blockdiagonal'
    elif args.pattern == 'R':
        args.pattern = 'blockrandom'

    generate_matrix(args.output, args.nrows, args.ncols, args.sparsity,
                    args.type, args.heatmap, args.pattern,
                    args.blocksize, args.blockrows, args.blockcols)


if __name__ == "__main__":
    main()

